core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -822,25 +822,42 @@ object Cause extends Serializable {
     final case class Filter[E](p: Cause[E] => Boolean) extends Folder[Any, E, Cause[E]] {
       def empty(context: Any): Cause[E] = Cause.empty
 
-      def failCase(context: Any, error: E, stackTrace: StackTrace): Cause[E] = Cause.Fail(error, stackTrace)
+      def failCase(context: Any, error: E, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Fail(error, stackTrace)
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
-      def dieCase(context: Any, t: Throwable, stackTrace: StackTrace): Cause[E] = Cause.Die(t, stackTrace)
+      def dieCase(context: Any, t: Throwable, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Die(t, stackTrace)
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
-      def interruptCase(context: Any, fiberId: FiberId, stackTrace: StackTrace): Cause[E] =
-        Cause.Interrupt(fiberId, stackTrace)
+      def interruptCase(context: Any, fiberId: FiberId, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Interrupt(fiberId, stackTrace)
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
       def bothCase(context: Any, left: Cause[E], right: Cause[E]): Cause[E] =
-        if (p(left)) {
-          if (p(right)) Cause.Both(left, right)
+        if (left.nonEmpty) {
```

Comment:
What's the rationale behind this change? Shouldn't we use the predicate when deciding if we're going to include left or right?

Reply:
I think the fact that are already filtered is because of how `bothCase` is used currently. I would be more comfortable keeping `p(c)` instead of assuming they're going to be filtered by the time we call this method.

Reply:
I actually think it's guaranteed by the fold semantics, furthermore applying `p(c)` won't be any cheaper that testing for emptiness.

Reply:
My point is that while it's guaranteed by the `fold` semantics, what's not guaranteed is that the method might be used in a different method. Since `Folder.Filter` is a public class (not sure why, but we can't ignore that it is), this test will fail using the current implementation:

```scala
          test("filter(false)") {
            val filter = Cause.Folder.Filter[Unit](_ => false)
            val cause = filter.bothCase((), f1, f2)
            assertTrue(cause.isEmpty)
          }
```

Reply:
I'm also not sure why the `Filter` class exists as a class instead of a `fold` call or anonymous class, nevertheless seems like it's being public is simply a miss-sight.
anyway, as a `Folder` it's intended to be used by `fold`s, I guess almost any other `Folder` implementation can be abused in similar ways to your example... what matters here is the semantics of filtering a `Cause` which are currently broken and have to be fixed, I think modifying the existing code is the path of least resistance, if you disagree we can always deprecate the exiting class and create a new one or simply replace it with a private/anonymous one.

Reply:
I think for now let's go with using `p(c)` instead of `nonEmpty` in `bothCase` and `thenCase`. I'm planning on looking into the performance of error paths in the future (including `Cause`), so maybe add a `TODO` comment so that we don't forget it